ml-conversational-analytic-tool/githubDataExtraction.py

```diff
@@ -1,179 +1,240 @@
+
+        try:
+            # sort by the first column: Pull request number
+            return pull_request_df.sort_values(by=pull_request_df.columns[0])
+        except IndexError:
+            return pull_request_df
 
-    def getAllPulls(self, reaction_flag=False):
+    def load_query(self, file_name, directory='queries'):
         """
-        Method to form a dataframe containing pull information. Parameters: name - name of exported csv file,
-        export - if the dataframe should be exported to csv. Returns: Dataframe with pull data
+        Method to load graphql queries
+        Parameters: file_name - name of the graphql query file
+        Returns: string
         """
-        self.reaction_flag = reaction_flag
-        if self.repo_opened:  # Verify if a repo has been opened
-            pull_data = []
-            pull_data.extend(self.getPullsByState('open'))  # Access all open pulls
-            pull_data.extend(self.getPullsByState('closed'))  # Access all closed pulls
-            return pd.DataFrame(pull_data)  # Return list of dictionaries converted to dataframe
-        print("Please open a Repo")
-
-    def getPullsByState(self, state):
+        directory_path = os.path.dirname(os.path.realpath(__file__))
+        query_path = os.path.join(directory_path, directory, file_name)
+
+        with open(query_path) as file:
+            return file.read()
+
+    def run_queries(self, query):
         """
-        Extract pulls with given state. Parameters: state - state of the pull (open or closed)
-        Return: list of dictionaries containing data regardining each pull
+        Method to run graphql queries
+        Parameters: query - dict of graphql query and variables
+        Returns: json string
         """
-        pull_data = []
-        try:  # Call the Github api to get all pulls
-            pulls = self.current_repo.get_pulls(state=state, sort='create')
-        except RateLimitExceededException as e:  # If token has reached limit
-            print("Token rate limit exceeded. Waiting for 1 hour", e)
-            time.sleep(60 * 60)  # Wait for 1 hour (time required to reset)
-            pulls = self.current_repo.get_pulls(state=state, sort='create')  # Get all pulls from Github API again
-        # Iterate over each pull
-        for pull in pulls:
-            try:  # Call to extract features for each pull
-                pull_data.append(self.getPullFeatures(pull))
-            except RateLimitExceededException as e:  # getPullFeatures accesses the Github API so provisino for rate limit
-                print("Token rate limit exceeded. Waiting for 1 hour", e)
-                time.sleep(60 * 60)
-                pull_data.append(self.getPullFeatures(pull))
-        return pull_data
-
-    def listOfComments(self, comments):
+        has_next_page = True
+        data = []
+
+        while has_next_page:
+            # wait two seconds to avoid the secondary rate limits
+            time.sleep(2)
```

Comment:
with my test, I rarely hit the secondary rate limit .. and adding 2 seconds on each iteration looks too much.

Reply:
I think the performance may depend on the machine of the user. Without any sleep, I run into 403 after a few calls have been made. Without waiting, it keeps making the call and does not resolve the 403 error.
```
 python ./ml-conversational-analytic-tool/githubDataExtraction.py kubernetes -R website  
retrieving data...
retrieving data...
retrieving data...
retrieving data...
retrieving data...
retrieving data...
retrieving data...
retrieving data...
retrieving data...
retrieving data...
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
```

I couldn't recreate the 403 error for when I set the wait time to 1 second, but I did run into it while testing yesterday. Therefore, decided to go with 2 seconds which I have not run into 403 error at all.

Since it's rarely thrown for 1 second, I'm happy to update to 1 second if 2 seconds is too long.

Reply:
UPDATE: I was able to recreate the 403 error with 1 second wait time using kubernetes/website 

```
retrieving data...
retrieving data...
retrieving data...
retrying due to 502 Bad Gateway {'data': None, 'errors': [{'message': 'Something went wrong while executing your query. This may be the result of a timeout, or it could be a GitHub bug. Please include `FCA9:69B8:55EC22:B9A6BF:620D4F00` when reporting this issue.'}]}
retrieving data...
retrieving data...
retrieving data...
retrieving data...
retrying due to 403 Forbidden {'documentation_url': 'https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits', 'message': 'You have exceeded a secondary rate limit. Please wait a few minutes before you try again.'}
retrieving data...
retrieving data...
retrieving data...
```

See that it runs into 403, but is able to resolve after waiting. Since it's pretty rare and it's able to resolve itself right away after waiting, I think updating to 1 second makes sense.

Reply:
Updated to 1 second